feat(core): time series ticks formatting

ilariaventurini opened the pull request:
### Updates

Closes: #299. Closes: #468.

Given a list of dates (the data to be represented), it recognizes the time interval necessary to represent the dates (15 seconds, minute, 30 minutes, hourly, daily (weekly), monthly, quarterly or yearly) and formats the ticks accordingly.

Two possible formats (primary and secondary) are associated with each time interval which can be used to format the current tick.

Below there is a short diagram:

    15seconds: 
      primary: "MMM d, pp"
      secondary: "pp"

      primary:
      * first tick
      * change day of month
      * change month
      * change year
      secondary:
      * otherwise

    minute: 
      primary: "MMM d, p"
      secondary: "p"

      primary:
      * first tick
      * change day of month
      * change month
      * change year
      secondary:
      * otherwise

    30minutes: 
      primary: "MMM d, p"
      secondary: "p"

      primary:
      * first tick
      * change day of month
      * change month
      * change year
      secondary:
      * otherwise

    hourly:
      primary: "MMM d, hh a"
      secondary: "hh a"

      primary:
      * first tick
      * change day of month
      * change month
      * change year
      secondary:
      * otherwise

    daily:
      primary: "MMM d"
      secondary: "d"

      primary:
      * first tick
      * change month
      * change year
      secondary:
      * otherwise

    weekly: 
      primary: "eee, MMM d"
      secondary: "eee"

      primary:
      * first tick
      * change week
      * change year
      secondary:
      * otherwise

    monthly: 
      primary: "MMM yyyy"
      secondary: "MMM"

      primary:
      * first tick
      * change year
      secondary:
      * otherwise

    quarterly:
      primary: "QQQ ''yy"
      secondary: "QQQ"

      primary:
      * first tick
      * first quarter
      secondary:
      * otherwise

    yearly: 
      primary: "yyyy"
      secondary: "yyyy"

      primary:
      * never
      secondary:
      * otherwise

Where:

- `MMM`: short name of the month (ie. `Jan-Dec`)
- `d`: day of the month (ie. `1-31`)
- `pp`: long localized time (ie. `12:00:00 AM`)
- `p`: long localized time (ie. `12:00 AM`)
- `hh`: hour [1-12] (ie. `01, 02, ..., 11, 12`)
- `a`: AM, PM
- `eee`: local day of week (ie. `Mon, Tue, Wed, ..., Su`)
- `yyyy`: calendar year (ie. `2010, 2017`)
- `yy`: calendar year (ie. `10, 17`)
- `QQQ`: quarter (ie. `Q1, ..., Q4`).

_Notes_:

- as you can see in the above diagram, the weekly time interval doesn't exist. It works as daily but it shows the day names instead of the day numbers and user can choose the format using the `showDayName` option

What we have done:

- [x] bold labels on primary ticks
- [x] formatting ticks using `date-fns`. This is the structure of `TimeScaleOptions`:

        interface TimeScaleOptions {
          addSpaceOnEdges?: number;
          showDayName?: boolean;
          timeIntervalFormats?: TimeIntervalFormats;
          localeCode?: string;
        }

        interface TimeIntervalFormats {
          "15seconds"?: TickFormats;
          "minute"?: TickFormats;
          "30minutes"?: TickFormats;
          "hourly"?: TickFormats;
          "daily"?: TickFormats;
          "weekly"?: TickFormats;
          "monthly"?: TickFormats;
          "quarterly"?: TickFormats;
          "yearly"?: TickFormats;
        }

        interface TickFormats {
          primary?: string;
          secondary?: string;
        }

`primary` and `secondary` specify the format of the formatted tick. The format of this string is based on [Unicode Technical Standard #35](https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table). 
More info on the [`date-fns` documentation](https://date-fns.org/v2.8.1/docs/format).

Each value is optional. Default values ​​allow you to build ticks similar to the [design](https://github.com/carbon-design-system/carbon-charts/issues/299).

As you can see, default locale code is `en-US` but the user can choose the one he prefers. The available codes are those supported by `date-fns`. [Here](https://github.com/date-fns/date-fns/tree/master/src/locale) is the list.

Basically by changing the locale only the language of the labels change, as seen in [this example](https://user-images.githubusercontent.com/44204353/74938415-400ae880-53ee-11ea-87d4-1ace68949741.png).
To change the formats of the labels according to a locale the only possibility is to add a new `timeIntervalFormats`. This is an open problem.

### Demo screenshots

![15seconds](https://user-images.githubusercontent.com/44204353/74938376-2c5f8200-53ee-11ea-8877-84a594dc7df1.png)
![minute](https://user-images.githubusercontent.com/44204353/74938386-2f5a7280-53ee-11ea-9698-8e4be1efc41c.png)
![30minutes](https://user-images.githubusercontent.com/44204353/74938389-31bccc80-53ee-11ea-8629-7edbcf69fdb6.png)
![hourly](https://user-images.githubusercontent.com/44204353/74938390-341f2680-53ee-11ea-9acf-5ec9db1c9d23.png)
![hourly custom](https://user-images.githubusercontent.com/44204353/74938395-371a1700-53ee-11ea-8a2d-af270fb9d593.png)
![daily](https://user-images.githubusercontent.com/44204353/74938402-397c7100-53ee-11ea-8189-8dad6048b829.png)
![weekly](https://user-images.githubusercontent.com/44204353/74938406-3bdecb00-53ee-11ea-9dc8-1911994b3400.png)
![monthly](https://user-images.githubusercontent.com/44204353/74938411-3da88e80-53ee-11ea-94fc-0b013ed85190.png)
![monthly custom](https://user-images.githubusercontent.com/44204353/74938415-400ae880-53ee-11ea-87d4-1ace68949741.png)
![quarterly](https://user-images.githubusercontent.com/44204353/74938422-426d4280-53ee-11ea-896e-ff972214e2c4.png)
![yearly](https://user-images.githubusercontent.com/44204353/74938424-44370600-53ee-11ea-8b61-3dc2ea468cf7.png)

Updating data, also the ticks changes:

![15seconds update](https://user-images.githubusercontent.com/44204353/75898446-59725280-5e3a-11ea-9e27-5094cc70781f.gif)
![hourly custom update](https://user-images.githubusercontent.com/44204353/75898465-62632400-5e3a-11ea-9902-74152cc784da.gif)

This PR solves the problem described in the [issue #468](https://github.com/carbon-design-system/carbon-charts/issues/468). If the dataset has only one datum, we show the tick in the most detailed way possible:

![single datum](https://user-images.githubusercontent.com/44204353/75908065-bc1f1a80-5e49-11ea-9ff4-e5c487d83242.png)

Demo tests and unit tests are updated.

In the future we would like to refactor the code inside `src/services/time-series` to make it more agnostic from the outside.

Currently all the tests created are both in the storybook and in the demo page.
We are already working on a PR that will allow to show some tests only in the storybook in order to keep the demo page cleaner.
Approved that PR we will put the tests of this brach only in the storybook.

### Review checklist (for reviewers only)

- [ ] Demos all features
- [ ] Documented/annotated
- [ ] Matches UI/UX specs
- [ ] Meets the [code style guide](https://github.com/IBM/carbon-charts/wiki/Code-style-guide)
- [ ] Accessible
- [ ] Mobile first (responsive)
- [ ] RTL support (bidirectional text)
- [ ] Performant (limited bloat)

caesarsol commented:
Let me add some notes on this PR I co-authored a bit:
 - the diff includes #496 because the fix was required for this PR: we will remove it after #496 is merged
 - we are still working on the localization of the date/time formats
   - @theiliad suggested not to maintain a list of date/time formats: we researched a bit and this is a problem, since the two formats usually given in the time libraries (date-fns, d3.time, moment) are not enough to get near the complexity in the time axis design proposal #299 
 - we are still working on the bold labels
 - we still need to agree on the added configuration property names :)

ilariaventurini commented:
Hi! Updated the PR description

theiliad commented:
> Hi! Updated the PR description

Cool, looking great so far 💯 

Wrt removing the `defaultLocale` code, as long as you can allow translating locale to a language then back to default once the `locale` has been removed from options, then it should be good

ilariaventurini commented:
Hi, I updated the PR description and changed the status from draft to ready for review.

ilariaventurini commented:
Have you had the time to look at this?

theiliad commented on packages/core/src/services/scales-cartesian.ts:
```diff
@@ -270,7 +269,10 @@ export class CartesianScales extends Service {
+			const spaceToAddToEdges = Tools.getProperty(options, "axes", axisPosition, "timeScale", "addSpaceOnEdges") !== null
+				? Tools.getProperty(options, "axes", axisPosition, "timeScale", "addSpaceOnEdges")
+				: Tools.getProperty(options, "timeScale", "addSpaceOnEdges");
+
```
If we're going to support per-axis timescale options then it makes sense to get rid of the global timeScale field, although we should probably deprecate it and warn the user.

ilariaventurini replied:
OK, working on it

theiliad replied:
@ilariaventurini let me know if you think that we should keep both

ilariaventurini replied:
Hi, no we think that only the global `timeScale` is necessary. In line 273 of `scales-cartesian.ts` there was an error of mine, which I have removed.
So the structure of the options is:
```
  options = {
    width,
    height,
    axes,
    timeScale: {...},
    ...
  }
```
Since the `timeScale` option has always been out of `axes`, there is no need for any warning because the structure has not changed and consequently nothing breaks :)

theiliad replied:
Great thanks Ilaria 🙂

theiliad commented on packages/core/src/services/time-series.ts:
```diff
@@ -0,0 +1,132 @@
+	const isFirstQuarter = Number(format((new Date(tick)), "q")) === 1;
+
+	switch (interval) {
+		case "15seconds": return (isFirstTick || isDayOfMonthChanged(tick) || isMonthChanged(tick) || isYearChanged(tick));
```
```suggestion
		case "15seconds":
			return (isFirstTick || isDayOfMonthChanged(tick) || isMonthChanged(tick) || isYearChanged(tick));
```

theiliad commented on packages/core/src/services/time-series.ts:
```diff
@@ -0,0 +1,132 @@
+		case "monthly": return (isFirstTick || isYearChanged(tick));
+		case "quarterly": return (isFirstTick || isFirstQuarter);
+		case "yearly": return false;
+		default: throw new Error(`${interval} is not a valid time interval.`);
```
Same indentations as the suggestion above

theiliad commented on packages/core/src/services/time-series.ts:
```diff
@@ -0,0 +1,132 @@
+function isYearChanged(timestamp: number): boolean {
+	const { M, d, s, m, H } = timestampToFormatTime(timestamp);
+	return M === 1 && d === 1 && H === 0 && m === 0 && s === 0;
+}
```
```suggestion
}

```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		left: { secondary: true },
+		bottom: {
+			scaleType: "time",
+			primary: true,
```
```suggestion
			primary: true
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		bottom: {
+			scaleType: "time",
+			primary: true,
+		},
```
```suggestion
		}
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		},
+	},
+	timeScale: {
+		localeCode: "fr",
```
```suggestion
		localeCode: "fr"
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+	},
+	timeScale: {
+		localeCode: "fr",
+	},
```
```suggestion
	}
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+				{ date: new Date(2017, 9), value: 10 },
+				{ date: new Date(2018, 0), value: 10 },
+				{ date: new Date(2018, 3), value: 10 },
+				{ date: new Date(2018, 6), value: 10 },
```
```suggestion
				{ date: new Date(2018, 6), value: 10 }
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		left: { secondary: true },
+		bottom: {
+			scaleType: "time",
+			primary: true,
```
```suggestion
			primary: true
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		bottom: {
+			scaleType: "time",
+			primary: true,
+		},
```
```suggestion
		}
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+			scaleType: "time",
+			primary: true,
+		},
+	},
```
```suggestion
	}
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+				{ date: new Date(1980, 0), value: 10 },
+				{ date: new Date(1981, 0), value: 10 },
+				{ date: new Date(1982, 0), value: 10 },
+				{ date: new Date(1983, 0), value: 10 },
```
```suggestion
				{ date: new Date(1983, 0), value: 10 }
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		left: { secondary: true },
+		bottom: {
+			scaleType: "time",
+			primary: true,
```
```suggestion
			primary: true
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		bottom: {
+			scaleType: "time",
+			primary: true,
+		},
```
```suggestion
		}
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+			scaleType: "time",
+			primary: true,
+		},
+	},
```
```suggestion
	}
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		{
+			label: "Dataset 1",
+			data: [
+				{ date: new Date("2020-01-06T19:43:16Z"), value: 10 },
```
```suggestion
				{ date: new Date("2020-01-06T19:43:16Z"), value: 10 }
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		left: { secondary: true },
+		bottom: {
+			scaleType: "time",
+			primary: true,
```
```suggestion
			primary: true
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+		bottom: {
+			scaleType: "time",
+			primary: true,
+		},
```
```suggestion
		}
```

theiliad commented on packages/core/demo/demo-data/time-series-axis.ts:
```diff
@@ -0,0 +1,333 @@
+			scaleType: "time",
+			primary: true,
+		},
+	},
```
```suggestion
	}
```

theiliad commented on packages/core/src/configuration.ts:
```diff
@@ -287,7 +301,7 @@ export const transitions = {
 	ticks: {
 		number: 7,
-		rotateIfSmallerThan: 30
+		rotateIfSmallerThan: 30,
```
```suggestion
		rotateIfSmallerThan: 30
```

theiliad commented on packages/core/src/configuration.ts:
```diff
@@ -121,7 +135,7 @@ const axisChart: AxisChartOptions = Tools.merge({}, chart, {
 	timeScale,
 	grid,
-	tooltip: axisChartTooltip
+	tooltip: axisChartTooltip,
```
```suggestion
	tooltip: axisChartTooltip
```

theiliad commented on packages/core/src/configuration.ts:
```diff
@@ -94,10 +94,24 @@ export const barChartTooltip: BarTooltipOptions = Tools.merge({}, axisChartToolt
+		"weekly": { primary: "eee, MMM d", secondary: "eee" },
+		"monthly": { primary: "MMM yyyy", secondary: "MMM" },
+		"quarterly": { primary: "QQQ ''yy", secondary: "QQQ" },
+		"yearly": { primary: "yyyy", secondary: "yyyy" },
```
```suggestion
		"yearly": { primary: "yyyy", secondary: "yyyy" }
```

theiliad commented on packages/core/src/interfaces/axis-scales.ts:
```diff
@@ -56,6 +56,45 @@ export interface AxisOptions {
+	 */
+	timeIntervalFormats?: TimeIntervalFormats;
+	/**
+	 * local code, ie. "en-US"
```
can we add a link here to the possible locale codes please? Maybe an upkg link to the locales folder in `fns` or if they have a page with the list that'd be great

theiliad commented on packages/core/src/interfaces/axis-scales.ts:
```diff
@@ -56,6 +56,45 @@ export interface AxisOptions {
+	primary?: string;
+	secondary?: string;
+}
+export interface TimeIntervalFormats {
```
```suggestion

export interface TimeIntervalFormats {
```

theiliad commented on packages/core/src/services/time-series.ts:
```diff
@@ -0,0 +1,132 @@
+
+// Return the formatted current tick
+export function formatTick(tick: number, i: number, interval: string, timeScaleOptions: TimeScaleOptions): string {
+	const showDayName = timeScaleOptions.showDayName;
```
Are we going to support per-axis timescale options or the global `timeScale` field?

Is there a better/cleaner way? Are we breaking anything?

theiliad commented:
There also seems to be a conflict now with the core storybook merge

gvnmagni commented:
Russel, this image shows an example of the solution we developed. Up in the conversation you can see more images and examples if you need them

![74938376-2c5f8200-53ee-11ea-8877-84a594dc7df1](https://user-images.githubusercontent.com/18484576/76978514-1682a580-6937-11ea-889d-bf03914bce59.png)

ilariaventurini commented:
Hi all, I updated the PR:

- solved conflicts
- I modified the code about the time domain expansion: I added the cases in which the difference between `endDate` and` startDate` is > 30 minutes, > 1 second, > 15 seconds. In this way we reflect the time intervals relating to the ticks formatting
- solved a problem in `all.stories` that did not allow you to view all the examples.

theiliad commented on packages/core/src/services/time-series.ts:
```diff
@@ -0,0 +1,140 @@
+import { format } from "date-fns";
+import * as locales from "date-fns/locale";
```
This will add to the bundle size of the library. If we don't directly use this piece we should take it out and switch

`throw new Error('Locale with code ${localeCode} not found. Avaible codes are: ${codes}.');` to just include a link to a page that lists all the codes

ilariaventurini replied:
Ok, I'll work on it!

theiliad commented on packages/core/src/services/time-series.ts:
```diff
@@ -0,0 +1,140 @@
+
+// Given a timestamp, return an object of useful time formats
+// Use Unicode date field symbol (https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table)
+export function timestampToFormatTime(timestamp: number) {
```
Can we think of a more readable name here?
I like your description in the comments above.

Maybe `getTimeformats(timestamp: number)`?

ilariaventurini replied:
Sure, you're right.

theiliad commented on packages/core/src/services/time-series.ts:
```diff
@@ -0,0 +1,140 @@
+}
+
+// Find the differences between consecutive numbers in an array
+function consecutiveDifferences(elements: number[]): number[] {
```
`findConsecutiveDifferences` or `getConsecutiveDifferences`?

natashadecoste commented on packages/core/demo/data/index.ts:
```diff
@@ -231,6 +232,71 @@ let allDemoGroups = [
+				data: timeSeriesAxisDemos.lineTimeSeriesDataSingleDatum,
+				options: timeSeriesAxisDemos.lineTimeSeriesSingleDatumOptions,
+				chartType: chartTypes.LineChart
+			},
```
```suggestion
			}
```

natashadecoste commented on packages/core/demo/data/index.ts:
```diff
@@ -184,7 +185,7 @@ let allDemoGroups = [
 				data: lineDemos.lineTimeSeriesDataRotatedTicks,
 				chartType: chartTypes.LineChart
-			}
+			},
```
```suggestion
			}
```

natashadecoste commented on packages/core/src/services/time-series.ts:
```diff
@@ -0,0 +1,124 @@
+		d: date.getDate(), // day of the month: 1-31
+		H: date.getHours(), // 24-hour clock: 0-23
+		m: date.getMinutes(), // minute: 0-59
+		s: date.getSeconds(), // seconds: 0-59
```
```suggestion
		s: date.getSeconds() // seconds: 0-59
```

ilariaventurini commented:
@theiliad Now it should work. A `Locale` was missing
